File: GmApps/src/app/message/message.service.ts
import { HttpClient } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { Paged } from '../models/Paged';
import { SendMessageRequest } from '../models/SendMessageRequest';

@Injectable({
  providedIn: 'root'
})
export class MessageService {

  constructor(private http: HttpClient) {
  }

  inbox(page: number = 1, unreadOnly: boolean = false): Observable<Paged<MessageDisplay>> {
    return this.http.get<Paged<MessageDisplay>>("/api/Message/Inbox", { params: { pagenum: String(page), unreadOnly: String(unreadOnly) }, withCredentials: true });
  }

  read(id: number, markRead: boolean = true): Observable<MessageDetails> {
    return this.http.get<MessageDetails>("/api/Message/Content", { params: { id: String(id), markRead: String(markRead) }, withCredentials: true });
  }

  outbox(page: number = 1): Observable<Paged<MessageDisplay>> {
    return this.http.get<Paged<MessageDisplay>>("/api/Message/Outbox", { params: { pagenum: String(page) }, withCredentials: true });
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>("/api/Message/Delete", { params: { id: String(id) }, withCredentials: true });
  }

  send(message: SendMessageRequest): Observable<boolean> {
    return this.http.post("/api/Message/Send", message, { withCredentials: true, observe: "response" }).pipe(
      map(r => r.ok)
    );
  }
}

File: GmGard.Client/Services/MessageService.cs
using System.Net.Http.Json;
using GmGard.Client.Models;

namespace GmGard.Client.Services;

public class MessageService
{
    private readonly HttpClient _http;
    public MessageService(HttpClient http) => _http = http;

    public async Task<PagedResult<MessageDisplay>?> InboxAsync(int page = 1, bool unreadOnly = false)
    {
        try { return await _http.GetFromJsonAsync<PagedResult<MessageDisplay>>(ApiRoutes.Message.Inbox(page, unreadOnly)); }
        catch { return null; }
    }

    public async Task<PagedResult<MessageDisplay>?> OutboxAsync(int page = 1)
    {
        try { return await _http.GetFromJsonAsync<PagedResult<MessageDisplay>>(ApiRoutes.Message.Outbox(page)); }
        catch { return null; }
    }

    public async Task<MessageDetails?> ReadAsync(int id, bool markRead = false)
    {
        try { return await _http.GetFromJsonAsync<MessageDetails>(ApiRoutes.Message.Content(id, markRead)); }
        catch { return null; }
    }

    public async Task<bool> DeleteAsync(int id)
    {
        try { var r = await _http.DeleteAsync(ApiRoutes.Message.Delete(id)); return r.IsSuccessStatusCode; }
        catch { return false; }
    }

    public async Task<(bool ok, string? error)> SendAsync(SendMessageRequest req)
    {
        try
        {
            var resp = await _http.PostAsJsonAsync(ApiRoutes.Message.Send, req);
            if (resp.IsSuccessStatusCode) return (true, null);
            var body = await resp.Content.ReadAsStringAsync();
            return (false, body);
        }
        catch (Exception ex) { return (false, ex.Message); }
    }

    public async Task<List<UserSuggestion>> SuggestUserAsync(string name)
    {
        if (string.IsNullOrWhiteSpace(name) || name.Length < 2) return [];
        try
        {
            var result = await _http.GetFromJsonAsync<List<UserSuggestion>>(ApiRoutes.Account.SuggestUser(name));
            return result ?? [];
        }
        catch { return []; }
    }
}
